drivers/gpio/gpio_mcp23x17.c

```diff
@@ -103,6 +126,33 @@ static int write_port_regs(const struct device *dev, uint8_t reg,
 	struct mcp23x17_drv_data *const drv_data =
 		(struct mcp23x17_drv_data *const)dev->data;
 	int ret;
+
```

Comment:
Your code here is assuming a little-endian system; it lacks the byte swapping done in the SPI version (`sys_cpu_to_le16`).  You don't need the union here, just do what the SPI version does.   Like below.

```
diff --git a/drivers/gpio/gpio_mcp23x17.c b/drivers/gpio/gpio_mcp23x17.c
index f44cb9f7e9..d57ad05af9 100644
--- a/drivers/gpio/gpio_mcp23x17.c
+++ b/drivers/gpio/gpio_mcp23x17.c
@@ -126,17 +126,22 @@ static int write_port_regs(const struct device *dev, uint8_t reg,
 	struct mcp23x17_drv_data *const drv_data =
 		(struct mcp23x17_drv_data *const)dev->data;
 	int ret;
+	uint16_t port_data;
+
+	LOG_DBG("MCP23X17: Write: REG[0x%X] = 0x%X, REG[0x%X] = 0x%X",
+		reg, (value & 0xFF), (reg + 1), (value >> 8));
+	port_data = sys_cpu_to_le16(value);
 
 #if MCP23X17_BUS_I2C
 	const struct mcp23x17_config *const config = dev->config;
 	const struct device *const i2c_master = drv_data->i2c_master;
 	uint16_t i2c_addr = config->i2c_slave_addr;
-	union gpio_mcp23x17_port_data *buffer = (union gpio_mcp23x17_port_data *)&value;
 	int retry = 3;
 
 	ret = -1;
 	while (retry--) {
-		ret = i2c_burst_write(i2c_master, i2c_addr, reg, buffer->byte, 2);
+		ret = i2c_burst_write(i2c_master, i2c_addr, reg,
+				      (uint8_t*)&port_data, sizeof(port_data));
 		if (ret == 0) {
 			break;
 		}
@@ -153,13 +158,6 @@ static int write_port_regs(const struct device *dev, uint8_t reg,
 #endif /* MCP23X17_BUS_I2C */
 
 #if MCP23X17_BUS_SPI
-	uint16_t port_data;
-
-	LOG_DBG("MCP23X17: Write: REG[0x%X] = 0x%X, REG[0x%X] = 0x%X",
-		reg, (value & 0xFF), (reg + 1), (value >> 8));
-
-	port_data = sys_cpu_to_le16(value);
-
 	uint8_t addr = MCP23X17_ADDR;
 	uint8_t buffer_tx[2] = { addr, reg };
 
@@ -169,8 +167,8 @@ static int write_port_regs(const struct device *dev, uint8_t reg,
 			.len = 2,
 		},
 		{
-			.buf = (uint8_t *)&port_data,
-			.len = 2,
+			.buf = &port_data,
+			.len = sizeof(port_data),
 		}
 	};
 	const struct spi_buf_set tx = {
```

Reply:
Thnx,
I'll look into this today when I've fixed the build.